### ISSUE TITLE
Windows Server Configuration for Drupal 7 WxT

### ISSUE THREAD

**Issue by mgifford** [wet-boew/wet-boew-drupal#377]
I'm a Linux guy, but I'd like to have the option of running this distro on Windows Server.  I heard that there was a performance comparison done between Windows & Linux using this distro & Windows won.

So how was Windows set up to perform so well?  What are versions were used and can the instructions for building this Windows server be found somewhere so that it is easier to replicate it.

Our client's got a very powerful server, but it loads Drupal super slowly.  Average page load times are over 12 seconds and there is no traffic on it to speak of.  

I'm just not sure where to send them for instructions on how to set up Windows correctly.

**Comment by sylus** [comment]
Apparently  tests have been done and performance is:

NGINX > IIS 7 > Apache

That being said IIS 7 does require a few certain tweaks to make everything run. Don't have a documentation on this now. Perhaps after the release date I could look at writing one but don't have the time right now.

**Comment by mgifford** [comment]
Just for clarification (as NGINX, IIS 7 & Apache can all run in Windows) is the fasted:
  http://nginx.org/en/docs/windows.html

And ya, we'd need to know what tweaks need to be made to IIS 7.

**Comment by joejoseph00** [comment]
Sounds like Nginx is the way to go.  Although we're running Apache and getting good performance out of it in production.  However we have very low traffic on our site.

**Comment by sylus** [comment]
Is this still a problem, IIS7 is pretty darn fast with few optimizations needed, I think the steps are pretty close to this.

http://chrislarson.me/content/chris-larson/installing-drupal-iis-microsoft-sql

**Comment by RobJohnston** [comment]
Were running a non-WxT Drupal installation in production on IIS6 without any such problems.  It has tens of thousands of pages and gets thousands of hits a day.  I've got IIS7 installed locally for development.  I gave the link above a cursory read and it seems to be accurate.  The only delay on loading a page happens when the server has been sitting idle for too long and the first page request runs the cron job.  Normally, it's zippy -- a 12 second page delay isn't normal.  I only see such a thing when I fire up my development box first thing in the morning.  No magical configuration options were done here for Drupal on Windows.

**Comment by mgifford** [comment]
@RobJohnston I'd love to get some more details about your config.  We get people asking about Drupal in Windows & just haven't been able to recommend it in the past.  If you've got a formula we can distribute it would really help.

Nice to see Drush coming along for Windows too:
http://www.ross.ws/content/drush-5-installation-and-basics-windows

Though not sure how to optimize it.  Are you also running MSSQL?

**Comment by RobJohnston** [comment]
Yes, IIS6 with SQL Server 2008 R2 in production and IIS7 with SQL Server Express 2008 for my local development.  There have been a few weird things with a fully MicroSoft set-up, like manually creating the sites/default/files/tmp directory.  There was also a patch for the sqlsrv module to ignore the length of integer fields (http://drupal.org/node/1258334).  The built-in archive view doesn't work (or any view with a contextual filter including a date), but I've got it working and may submit the patch in the coming weeks.  There's a query that cron runs that can bomb after a large content change (http://drupal.org/node/1763914).  A few other things to get deploy and services modules working.  Obscure things like the quiz module need a lot of work.  

I've been thinking of writing a batch file to do all the Windows/SQL Server specific installation steps (e.g. using icacls to set permissions and maybe sqlcmd to do some database work), as in issue #400, but I haven't gotten around to it.

As you can see, most problems are with SQL Server.  All in all, everything is possible but might require applying patches that others don't need.  

I started with "Installing Druapal on Windows" (http://drupal.org/node/970492).  Maybe the "Drupal on Windows" group would help too (http://groups.drupal.org/drupal-windows).  Good Drush link... thanks for that.

**Comment by mgifford** [comment]
I pushed a bit on a couple of the issues you listed.  Would be great to have more folks test this.  Do you have any specs on the server you're using?

**Comment by RobJohnston** [comment]
The web servers are running IIS6 with PHP 5.3 running on FastCGI 1.5.  Ionic ISAPI Rewriting Filter is used to make friendly URLs.  Physically, it's got 32 GB of ram and using about 10%, around 40-some 1.87 GHz Xeon dual-core CPUs and using 1%.... and there are 3 machines like this in load balance!  A bit too much.

**Comment by RobJohnston** [comment]
@mgifford ... FYI:  Last month's numbers: > 3 million visits, > 4.3 million views,  >1.5 million unique visitors, and > 41.8 million hits.  No 12 second load times here!

**Comment by mgifford** [comment]
Nice.  What's the URL?

**Comment by RobJohnston** [comment]
Sorry @mgifford, but I can't say.  We can use Drupal and contribute back to the project on the condition that we don't tell that the site is running Drupal now (I know the problem with that...not my decision).

**Comment by mgifford** [comment]
With a Google search I'm pretty certain I know. Let's just say I'm happy to see such a prominent site be using Drupal & testing it (successfully) with a Windows platform.  I'll refrain from listing said sites on great lists like http://groups.drupal.org/government-sites#Canada but do hope that those who make these decisions change their minds.

**Comment by mgifford** [comment]
@RobJohnston If there's anything else you can contribute back to the documentation for Windows configurations, that would be great.  Thanks.

**Comment by RobJohnston** [comment]
I can tell you that we upgraded to Windows Server 2008 R2 (which runs IIS 7.5).  This allowed us to dump Ionic ISAPI Rewriting Filter and use the web.config file instead.  It also let us install the WinCache Extension for PHP (caches PHP bytecode in memory), which seems a bit faster.  Upgrades also include SQL Server 2008 R2 and PHP 5.4.  I forget the reason for the Web server upgrade, but it was something Drupal related and many more of the tests pass now.  Everyone else seemed happy to have an excuse to upgrade to software that's only 5 years old :-).

On the testing machine, which is virtual, we've noticed that occasionally things do slow down but haven't investigated to determine why.  There is still that first-load-in-the-morning slow down as the cron job is running, which is triggered by the first page request, but usually nothing too horrid after that.
